src/schnorrsig.rs

```diff
@@ -0,0 +1,435 @@
+            }
+            _ => Err(Error::InvalidSchnorrNonce),
+        }
+    }
+}
+
+impl From<ffi::SchnorrNonce> for SchnorrNonce {
+    #[inline]
+    fn from(nonce: ffi::SchnorrNonce) -> SchnorrNonce {
+        SchnorrNonce(nonce)
+    }
+}
+
+impl From<SecretKey> for SchnorrNonce {
+    #[inline]
+    fn from(sk: SecretKey) -> SchnorrNonce {
+        SchnorrNonce::from_slice(&sk[0..constants::SCHNORR_NONCE_SIZE]).unwrap()
+    }
+}
+
+impl From<PublicKey> for SchnorrNonce {
+    #[inline]
+    fn from(pubkey: PublicKey) -> SchnorrNonce {
+        SchnorrNonce::from_slice(&pubkey.serialize()[1..33]).unwrap()
+    }
+}
+
+impl CPtr for SchnorrNonce {
+    type Target = ffi::SchnorrNonce;
+    fn as_c_ptr(&self) -> *const Self::Target {
+        self.as_ptr()
+    }
+
+    fn as_mut_c_ptr(&mut self) -> *mut Self::Target {
+        self.as_mut_ptr()
+    }
+}
+
+impl SchnorrNonce {
+    /// Creates an SchnorrNonce directly from a slice
+    #[inline]
+    pub fn from_slice(data: &[u8]) -> Result<SchnorrNonce, Error> {
+        match data.len() {
+            constants::SCHNORR_NONCE_SIZE => {
+                let mut ret = [0; constants::SCHNORR_NONCE_SIZE];
+                ret[..].copy_from_slice(data);
+                Ok(SchnorrNonce(ffi::SchnorrNonce::from_array(ret)))
+            }
+            _ => Err(Error::InvalidSchnorrNonce),
+        }
+    }
+
+    /// Obtains a raw const pointer suitable for use with FFI functions
+    #[inline]
+    pub fn as_ptr(&self) -> *const ffi::SchnorrNonce {
+        &self.0 as *const _
+    }
+
+    /// Obtains a raw mutable pointer suitable for use with FFI functions
+    #[inline]
+    pub fn as_mut_ptr(&mut self) -> *mut ffi::SchnorrNonce {
+        &mut self.0 as *mut _
+    }
+}
+
+impl<C: Signing> Secp256k1<C> {
+    /// Create a schnorr signature.
+    pub fn schnorr_sign(
+        &self,
+        msg: &Message,
+        sk: &SecretKey,
+        aux_rand: &[u8; 32],
+    ) -> Result<SchnorrSignature, Error> {
+        unsafe {
+            let mut keypair = ffi::KeyPair::new();
+            let ret = ffi::secp256k1_keypair_create(self.ctx, &mut keypair, sk.as_c_ptr());
+            if ret == 0 {
+                return Err(Error::InvalidSecretKey);
+            }
+
+            let mut sig = ffi::SchnorrSignature::new();
+            assert_eq!(
+                1,
+                ffi::secp256k1_schnorrsig_sign(
+                    self.ctx,
+                    &mut sig,
+                    msg.as_c_ptr(),
+                    &keypair,
+                    ffi::secp256k1_nonce_function_bip340,
+                    aux_rand.as_c_ptr() as *const ffi::types::c_void
+                )
+            );
+
+            Ok(SchnorrSignature(sig))
+        }
+    }
+
+    /// Create a schnorr signature using the provided nonce.
+    pub fn schnorr_sign_with_nonce(
+        &self,
+        msg: &Message,
+        sk: &SecretKey,
+        nonce: &SchnorrNonce,
```

Comment:
Actually I think this is meant to be a private key, `k`, whereas the type `SchnorrNonce` represents public keys `R = k*G`. The name is kind of confusing because both `k` and `R` values are sometimes called the nonce...

Reply:
Good catch thanks! I indeed got confused because it was called nonce, changed to private key.